### PR TITLE
Cancel swap

### PR DESCRIPTION
**Pull request by TheCharlatan**
Based on #499 and #485 .

A swap should be cancelable by the user if its side has not locked any coins yet.
    
For Alice, this means a swap can be canceled without additional handling up until she reaches state RefundSigA. Once she reaches RefundSigA, the XMR funding address is displayed and may thus contain locked coins. At the point, the swap should just follow the rest of  the protocol. 
    
For Bob, this means a swap can be canceled without additional handling as long as it still is in its original start state. If it is in commit, reveal, or corearb state and has not yet received the refund procedure signatures from Alice, it additionally attempts to sweep any coins that were sent to the funding address.

~Because of the sweeping, the syncers need to keep on running after the user has canceled and the swap is already being cleaned up.~

### CODE REVIEW

File: src/swapd/runtime.rs [hunk 1]
```diff
@@ -2056,8 +2087,6 @@ impl Runtime {
                     }),
                 )?;
 
-                self.send_peer(endpoints, Msg::RefundProcedureSignatures(refund_proc_sigs))?;
```
Comment (TheCharlatan) [inline src/swapd/runtime.rs:1828]: Undo this change, this handled correctly as it is.

**Review by zkao — CHANGES_REQUESTED** [review]
The sequence of messages seems suboptimal, and that created extra complications, it looks straightforward for me to fix it if so wished

File: src/swapd/runtime.rs [hunk 2]
```diff
@@ -2005,6 +2023,33 @@ impl Runtime {
...
+                    Request::SyncerTask(task),
+                )?;
+
+                if self.awaiting_btc_sweep {
+                    if let Some((source, request)) =
+                        self.pending_requests.remove_entry(&ServiceId::Farcasterd)
+                    {
+                        self.handle_rpc_ctl(endpoints, source, request[0].request.clone())?;
+                    }
+                    self.awaiting_btc_sweep = false;
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:2049]: How so?

File: src/swapd/syncer_client.rs [hunk 3]
```diff
@@ -267,6 +268,18 @@ impl SyncerState {
...
+        let id = self.tasks.new_taskid();
+        let lifetime = self.task_lifetime(Coin::Bitcoin);
+        let addendum = SweepAddressAddendum::Bitcoin(sweep_bitcoin_address);
+        let sweep_task = SweepAddress {
+            id,
+            lifetime,
+            addendum,
+            from_height: None,
+        };
+        Task::SweepAddress(sweep_task)
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/syncer_client.rs:281]: Will do.

File: src/farcasterd/runtime.rs [hunk 4]
```diff
@@ -1507,10 +1526,23 @@ impl Runtime {
...
+                        }
+                    }
+                    Coin::Monero => {
+                        if self.funding_xmr.remove(&get_swap_id(&source)?).is_some() {
+                            self.stats.incr_funding_monero_canceled();
+                            true
+                        } else {
+                            false
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/farcasterd/runtime.rs:1544]: This was a bad copy-pasta :( - will fix.

**Comment by TheCharlatan** [comment]
Thanks for the review @zkao . We should discuss the sweeping a bit more. The problem is that the sweep btc task never emits an event if the user never funded. I will add a one-shot flag to the sweep btc task to only try once and even no funds are found and sweeped, emit an event reporting the sweep failure to swapd. Then we can consume the event in swapd and trigger the clean up as you proposed.

File: src/swapd/runtime.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zkao) [reply src/swapd/runtime.rs:2049]: on the recursion swapd terminates, before setting  await_btc_sweep to false

[the remark this answers is not in the document]
Comment (zkao) [reply src/swapd/runtime.rs:2049]: not infinite loop, but it goes further than it should

[the remark this answers is not in the document]
Comment (zkao) [reply src/swapd/runtime.rs:2049]: self.awaiting_btc_sweep = false;

before hand_rpc_ctl should do it

File: src/swapd/runtime.rs [hunk 5]
```diff
@@ -2005,6 +2023,33 @@ impl Runtime {
...
+                let task = self.syncer_state.sweep_btc(sweep_bitcoin_address);
+
+                endpoints.send_to(
+                    ServiceBus::Ctl,
+                    self.identity(),
+                    self.syncer_state.bitcoin_syncer(),
+                    Request::SyncerTask(task),
+                )?;
+
+                if self.awaiting_btc_sweep {
```
[the remark this answers is not in the document]
Comment (zkao) [reply src/swapd/runtime.rs:2043]: I'm not convinced so much complication justifies handling a seemingly trivial special case

[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/swapd/runtime.rs:2043]: Thanks for the feedback, I'll kick this back to draft and rework the message flow into something more sane.

**Comment by TheCharlatan** [comment]
I reworked the PR and was able to remove most quirks of the workflow and any new state in the runtimes it introduced. Sweeping bitcoin is now a one-shot attempt and won't be continuously re-attempted until its lifetime expires.

File: src/swapd/runtime.rs [hunk 6]
```diff
@@ -2105,6 +2144,66 @@ impl Runtime {
...
+                    endpoints,
+                    source,
+                    Request::String("Canceled swap".to_string()),
+                )?;
+            }
+            Request::CancelSwap
+                if self.state.b_commit()
+                    || self.state.b_reveal()
+                    || (!self.state.b_received_refund_procedure_signatures()
+                        && self.state.b_core_arb()) =>
```
Comment (zkao) [inline src/swapd/runtime.rs:2176]: Wouldn't it be worth for readability to add a some btc funded field in those states?

File: src/syncerd/bitcoin_syncer.rs [hunk 7]
```diff
@@ -930,42 +930,41 @@ fn sweep_polling(
...
+            if !sweep_addresses.is_empty() {
+                match Client::new(&electrum_server) {
+                    Err(err) => {
+                        error!(
+                            "Failed to create btc sweep electrum client: {}, retrying",
+                            err
+                        );
+                    }
+                    Ok(client) => {
+                        for (id, sweep_address_task) in sweep_addresses.iter() {
```
Comment (zkao) [inline src/syncerd/bitcoin_syncer.rs:942]: i would put this for loop where `if !sweep_addresses.is_empty()` is, and remove the latter

**Review by zkao — APPROVED** [review]
Now it makes total sense for me, thank you, minor comments only that do not need to be addressed at this time

File: src/syncerd/bitcoin_syncer.rs (hunk 7 above)
Comment (TheCharlatan) [reply src/syncerd/bitcoin_syncer.rs:942]: The problem is that you are then creating a new electrum client on every iteration. This was a feedback from @h4sh3d when he reviewed the sweep PR.You'd only want to create one once when there are some addresses to be swept.

Comment (zkao) [reply src/syncerd/bitcoin_syncer.rs:942]: ACK
